CMPUT404F21TEAM/social-distribution
# Refactor author updates to use threads

Try to improve the response time of the web app by doing some caching and making updates via HTTP requests within separate threads. This code helps make the `/app/authors` page load much faster by performing all HTTP requests off of the main thread. The page will return to the user immediately, and the data models will be updated once the data is available.

On my machine
- `master` --> 40 seconds to load `/app/authors`
- `diego/refactor-update-fetch` --> less than one second to load, refresh 5 seconds later and all remote author data should be there (note this may take longer on Heroku)

### Summary:
 - created `fetchers.py` (yeah maybe could have chosen a better name) that contains a bunch of "fetch" functions. These functions will initiate an update process of some sort. 
 - Moved API calls out of the `Author` model. These calls are now only done in `fetchers.py`. This is to improve code cohesion but was also necessary to avoid cyclic dependency errors.
 - Follows now keep track of follows between all authors, not just local authors and authors. This is so that we can cache this data and access it immediately when rendering pages. We now check for friendship by seeing it two follows exists (bidirectional)
 - Only delete authors if we get a 404 and increase timeout for GETs
 - One limitation right now is that updates will only be fetched _after_ the data is rendered to the webpage. We either need ajax to get updated data once it is available, or we just accept this limitation and refresh the page. The idea is that data won't change too often and the user will be clicking between webpages anyways, in which case they will see the updated data.

### How to test:
 - Clone the branch, set `HOST = "social-distribution-fall2021.herokuapp.com"`, delete the DB, then migrate. This will trick the migration files into adding remote hosts to the Node table. Make an account then go to `/app/authors`. You should see the page return quickly with nothing, but after a few seconds refresh and all remote authors will show.
 - Delete DB again, comment out  `HOST = "social-distribution-fall2021.herokuapp.com"`, then migrate. Also, run the dummy server in another repo. Use this one to test out all regular features such as sending follow requests, accepting friend requests, removing followers, sending posts, etc.
 - Delete DB again, switch to master, migrate and create some authors and posts. Switch to the new branch and update the existing database with extra migrations. Make sure no errors occur as we don't want to reset the database on the deployment again.
